src/meta/expr.lean

```diff
@@ -307,14 +307,15 @@ namespace expr
 
 /--
 Turns an expression into a natural number, assuming it is only built up from
-`has_one.one`, `bit0`, `bit1`, `has_zero.zero`, `nat.zero`, and `nat.succ`.
+`has_one.one`, `bit0`, `bit1`, `has_zero.zero`, `nat.zero`, `nat.succ`, or `(+ 1)`.
 -/
 protected meta def to_nat : expr → option ℕ
 | `(has_zero.zero) := some 0
 | `(has_one.one) := some 1
 | `(bit0 %%e) := bit0 <$> e.to_nat
 | `(bit1 %%e) := bit1 <$> e.to_nat
 | `(nat.succ %%e) := (+1) <$> e.to_nat
+| `(%%e + 1) := (+1) <$> e.to_nat -- a different form for `nat.succ`
```

Comment:
This function is intended to identify if a number is a normalized numeral, which is why it doesn't have the whole gamut of numerical operations in it. Numbers built from these constructors will print as numerals, which is not true for `+1`. You should use `eval_expr` if you want to turn an arbitrary expression into a nat.